File: javascript/src/0001_TwoSum/TwoSumSolutionD.js
import { BaseTwoSumSolution } from './BaseTwoSumSolution.js';

// Brute Force Solution
export class TwoSumSolutionD extends BaseTwoSumSolution {
	// Time Complexity: O( N )
	// Space Complexity: O( N )
	twoSum(nums, target) {
		const count = nums.length;
		const map = new Map();

		// map all of the nums to their locations
		for (let i = 0; i < count; i++) {
			const num = nums[i];
			const complement = target - num;

			// but also check for complements along the way
			if (map.has(complement)) {
				return [i, map.get(complement)];
			}

			map.set(num, i);
		}

		throw new Error('No solution found.');
	}
}

File: csharp/src/Solutions.Lib/0001_TwoSum/TwoSumSolutionD.cs
namespace Solutions.Lib;

/// <summary>
/// A One-Pass Hash Table solution
/// </summary>
public class TwoSumSolutionD : TwoSumSolution
{
	// Space Complexity: O( N )
	// Time Complexity: O( N )
	protected override int[] TwoSum(int[] nums, int target)
	{
		int count = nums.Length;
		Dictionary<int, int> map = new();

		// map all of the nums to their locations
		for (int i = 0; i < count; i++)
		{
			int num = nums[i];
			int complement = target - num;

			// but also check for complements along the way
			if (map.ContainsKey(complement))
			{
				return new int[] { i, map[complement] };
			}

			map[num] = i;
		}

		throw new ArgumentException("No solution found.");
	}
}
